### DISCUSSION TITLE
What were variable GC handles used for?

### DISCUSSION THREAD

**Discussion by AustinWise** [dotnet/runtime#86840] [Q&A] [answered]
Specifically `HNDTYPE_VARIABLE`. The only place I see it referred to is CoreRT (`RhpHandleAllocVariable`), but it does not seem to be used there for anything. Perhaps .NET Native uses it?

**Comment by jkotas** [comment, marked as answer]
Yes, the variable GC handles were used for WinRT/COM interop in .NET Native for UWP for a period of time.

They were alternative of ref-counted handles. Ref-counted handles require GC to call back into COM interop subsystem to find out whether the handle is strong or not. Variable handles flip the direction. COM interop had to update the handle type every time when the ref count dropped to zero or when the ref-count was bumped from zero to one.

I do not recall the exact reason why we have abandoned them. The shipping NET Native for UWP does not use them anymore.
